### PR TITLE
Add vault auto auth for kubernetes

### PR DESCRIPTION
**Pull request by saheienko**
Signed-off-by: Serhii Aheienko <serhii.aheienko@gmail.com>

### CODE REVIEW

File: vault/vault.go [hunk 1]
```diff
@@ -79,7 +92,23 @@ func New(
...
+		}
+
+		autoAuth = true
+		client.SetToken(token)
+	} else {
+		token := getVaultParam(secretConfig, api.EnvVaultToken)
+		if token == "" {
+			return nil, ErrVaultTokenNotSet
+		}
+		client.SetToken(token)
```
Comment (adityadani) [inline vault/vault.go]: client.SetToken can be moved out of if and else.

File: vault/vault.go [hunk 2]
```diff
@@ -31,13 +43,19 @@ var (
...
 
 type vaultSecrets struct {
-	client        *api.Client
+	mu     sync.RWMutex
+	client *api.Client
+
 	endpoint      string
 	backendPath   string
 	isKvBackendV2 bool
+	autAuth       bool
```
Comment (adityadani) [inline vault/vault.go]: nit: rename to autoAuth

File: vault/vault.go [hunk 3]
```diff
@@ -250,6 +359,56 @@ func configureTLS(config *api.Config, secretConfig map[string]interface{}) error
...
+}
+
+func buildAuthConfig(config map[string]interface{}) (*auth.AuthConfig, error) {
+	role := getVaultParam(config, AuthKubernetesRole)
+	if role == "" {
+		return nil, ErrVaultAuthKubernetesRole
+	}
+	mountPath := getVaultParam(config, AuthMountPath)
+	if mountPath == "" {
+		mountPath = "auth/kubernetes"
```
Comment (adityadani) [inline vault/vault.go]: nit: make "auth/kubernetes" a constant

File: vault/vault.go [hunk 4]
```diff
@@ -250,6 +359,56 @@ func configureTLS(config *api.Config, secretConfig map[string]interface{}) error
...
+	if err != nil {
+		return "", err
+	}
+
+	path, data, err := method.Authenticate(context.TODO(), client)
+	if err != nil {
+		return "", err
+	}
+
+	secret, err := client.Logical().Write(path, data)
```
Comment (adityadani) [inline vault/vault.go]: do we need extra permissions now to perform this operation ?

File: vault/vault.go [hunk 5]
```diff
@@ -193,6 +225,79 @@ func (v *vaultSecrets) ListSecrets() ([]string, error) {
 	return nil, secrets.ErrNotSupported
 }
 
+func (v *vaultSecrets) read(path string) (*api.Secret, error) {
+	secretValue, err := v.lockedRead(path)
+	if v.isTokeExpired(err) {
+		if err = v.renewToken(); err != nil {
+			return nil, err
```
Comment (adityadani) [inline vault/vault.go]: lets return the original error here instead of the error received as part of renewToken

File: vault/vault.go [hunk 6]
```diff
@@ -193,6 +225,79 @@ func (v *vaultSecrets) ListSecrets() ([]string, error) {
...
+		}
+		return v.lockedRead(path)
+	}
+	return secretValue, err
+}
+
+func (v *vaultSecrets) write(path string, data map[string]interface{}) (*api.Secret, error) {
+	v.mu.RLock()
```
Comment (adityadani) [inline vault/vault.go]: no need of Rlock here as lockedWrite is anyways taking a Rlock

File: vault/vault.go (hunk 4 above)
Comment (saheienko) [reply vault/vault.go]: No, it doesn't require any additional rules to the portworx policy.

File: vault/vault.go (hunk 5 above)
Comment (saheienko) [reply vault/vault.go]: Updated.

File: vault/vault.go (hunk 6 above)
Comment (saheienko) [reply vault/vault.go]: Updated.

File: vault/vault.go (hunk 3 above)
Comment (saheienko) [reply vault/vault.go]: Updated.

File: vault/vault.go (hunk 2 above)
Comment (saheienko) [reply vault/vault.go]: Updated.

File: vault/vault.go (hunk 1 above)
Comment (saheienko) [reply vault/vault.go]: Updated.

**Pull request merged**